### PR TITLE
Alexandros.targets validation vignette

### PR DESCRIPTION
**Pull request by alexverse**
Tutorial for targets with two scenarios:

1. Metadata validation with a controlled exit
2. Error-localisation and data curation

If we add it as a regular vignette the html widgets should be captured as `png` for CRAN submission.

### CODE REVIEW

File: vignettes/targets_workflow.Rmd [hunk 1]
```diff
@@ -0,0 +1,540 @@
...
+}
+
+get_schema(input_data) %T>%
+  print() %>%
+  write_csv("schema.csv")
+```
+
+## Metadata validation
+
+```{r, include = FALSE, eval = TRUE}
```
Comment (jakubnowicki) [inline vignettes/targets_workflow.Rmd:102]: Please add a comment to all code chunks that are not included with a short description (what is happening and why it is needed).

File: vignettes/targets_workflow.Rmd [hunk 2]
```diff
@@ -0,0 +1,540 @@
...
+```{r, eval=FALSE}
+# R/functions.R
+get_data <- function(file) {
+  read_csv(file, col_types = cols())
+}
+
+metadata_validator <- function(data, schema) {
+  tar_assert_identical(names(schema), c("name", "type"))
+
+  report <- data_validation_report()
```
Comment (jakubnowicki) [inline vignettes/targets_workflow.Rmd:195]: Please add an empty line below.

File: vignettes/targets_workflow.Rmd [hunk 3]
```diff
@@ -0,0 +1,540 @@
...
+  validate(data) %>%
+    validate_if(
+      has_all_names(!!!schema$name),
+      description = "Columns exist"
+    ) %>%
+    add_results(report)
```
Comment (jakubnowicki) [inline vignettes/targets_workflow.Rmd:201]: Please add an empty line below.

File: vignettes/targets_workflow.Rmd [hunk 4]
```diff
@@ -0,0 +1,540 @@
...
+    var_names <- schema %>%
+      filter(type == col_class) %>%
+      pull(name)
+    validate(data) %>%
+      validate_if(
+        has_class(!!!var_names, class = {{ col_class }}),
+        description = glue::glue("Column classes: {col_class}")
+      ) %>%
+      add_results(report)
+  })
```
Comment (jakubnowicki) [inline vignettes/targets_workflow.Rmd:212]: Please add an empty line below.

File: vignettes/targets_workflow.Rmd [hunk 5]
```diff
@@ -0,0 +1,540 @@
...
+  walk(unique(schema$type), \(col_class) {
+    var_names <- schema %>%
+      filter(type == col_class) %>%
+      pull(name)
+    validate(data) %>%
+      validate_if(
+        has_class(!!!var_names, class = {{ col_class }}),
+        description = glue::glue("Column classes: {col_class}")
+      ) %>%
+      add_results(report)
```
Comment (jakubnowicki) [inline vignettes/targets_workflow.Rmd:211]: This is quite a lot for a lambda function, please consider extracting this part to a named function.

File: vignettes/targets_workflow.Rmd [hunk 6]
```diff
@@ -0,0 +1,540 @@
...
+  })
+  report
+}
+
+validation_violation <- function(report) {
+  violations_exist <- report$get_validations() %>%
+    summarise(
+      sum(num.violations, na.rm = TRUE) > 0
+    ) %>%
+    pull()
```
Comment (jakubnowicki) [inline vignettes/targets_workflow.Rmd:221]: Please add an empty line below.

File: vignettes/targets_workflow.Rmd [hunk 7]
```diff
@@ -0,0 +1,540 @@
...
+  if (isTRUE(violations_exist)) {
+    abort(
+      "Validation schema error",
+      body = capture.output(report),
+      class = "validation_violation"
+    )
+  }
```
Comment (jakubnowicki) [inline vignettes/targets_workflow.Rmd:228]: Please add an empty line below.

File: vignettes/targets_workflow.Rmd [hunk 8]
```diff
@@ -0,0 +1,540 @@
...
+
+    -   The column classes in the schema table comply to when we read the `data.csv` file from disk.
+
+-   `validation_violation()` checks for validation violations in the `Report` object returned by the `data.validator`. Following a defensive approach we *'fail fast'* in case of a violation, to protect downstream steps to be invalidated or produce an uncontrolled error. Furthermore, we are using an error with *classed condition* that can be handled selectively, allowing for fine-grained error handling.
+
+The `targets` script file for metadata validation looks like:
+
+```{r, eval=FALSE}
+# _targets.R file
+library(targets)
```
Comment (jakubnowicki) [inline vignettes/targets_workflow.Rmd:247]: Please add a blank line below.

File: vignettes/targets_workflow.Rmd [hunk 9]
```diff
@@ -0,0 +1,540 @@
...
+    -   All column names provided in the schema are present in the data.
+
+    -   The column classes in the schema table comply to when we read the `data.csv` file from disk.
+
+-   `validation_violation()` checks for validation violations in the `Report` object returned by the `data.validator`. Following a defensive approach we *'fail fast'* in case of a violation, to protect downstream steps to be invalidated or produce an uncontrolled error. Furthermore, we are using an error with *classed condition* that can be handled selectively, allowing for fine-grained error handling.
+
+The `targets` script file for metadata validation looks like:
+
+```{r, eval=FALSE}
+# _targets.R file
```
Comment (jakubnowicki) [inline vignettes/targets_workflow.Rmd:246]: Please add a blank line below.

File: vignettes/targets_workflow.Rmd [hunk 10]
```diff
@@ -0,0 +1,540 @@
...
+library(targets)
+source("R/functions.R")
+options(crayon.enabled = FALSE)
+tar_option_set(
+  packages = c("readr", "dplyr", "data.validator", "assertr", "rlang",
+               "purrr")
+)
```
Comment (jakubnowicki) [inline vignettes/targets_workflow.Rmd:253]: Please add a blank line below.

File: vignettes/targets_workflow.Rmd [hunk 11]
```diff
@@ -0,0 +1,540 @@
...
+    -   The column classes in the schema table comply to when we read the `data.csv` file from disk.
+
+-   `validation_violation()` checks for validation violations in the `Report` object returned by the `data.validator`. Following a defensive approach we *'fail fast'* in case of a violation, to protect downstream steps to be invalidated or produce an uncontrolled error. Furthermore, we are using an error with *classed condition* that can be handled selectively, allowing for fine-grained error handling.
+
+The `targets` script file for metadata validation looks like:
+
+```{r, eval=FALSE}
+# _targets.R file
+library(targets)
+source("R/functions.R")
```
Comment (jakubnowicki) [inline vignettes/targets_workflow.Rmd:248]: Please add a blank line below.

File: vignettes/targets_workflow.Rmd [hunk 12]
```diff
@@ -0,0 +1,540 @@
...
+
+This time running the pipeline completes successfully, and all unnecessary steps are skipped, which is incredibly time-efficient in large projects:
+
+```{r}
+tar_make_catch()
+```
+
+## Error-localization and data curation
+
+```{r, include = FALSE, eval = TRUE}
```
Comment (jakubnowicki) [inline vignettes/targets_workflow.Rmd:334]: Please add a description.

File: vignettes/targets_workflow.Rmd [hunk 13]
```diff
@@ -0,0 +1,540 @@
...
+5.  Check values against code list
+6.  Conditional checks between variables
+7.  Outlier detection
+
+Using the `data.validator::validate_col()` function we give an example of explicitly validating columns, for uniqueness, existence of particular levels, and outliers. When particular columns and rows are checked against some assertion the exact place where the validation failure occurred can be tracked down using `data.validator::get_results()`. With this information at hand it is straightforward to fix known data issues and keep track of the data health between consequent curation steps.
+
+The functions that compute the pipeline "targets" are as follows:
+
+```{r, eval=FALSE}
+# R/functions.R
```
Comment (jakubnowicki) [inline vignettes/targets_workflow.Rmd:429]: Please add a blank line below.

File: vignettes/targets_workflow.Rmd [hunk 14]
```diff
@@ -0,0 +1,540 @@
...
+get_data <- function(file) {
+  read_csv(file, col_types = cols())
+}
+
+columns_validator <- function(data) {
+  report <- data_validation_report()
```
Comment (jakubnowicki) [inline vignettes/targets_workflow.Rmd:435]: Please add a blank line below.

File: vignettes/targets_workflow.Rmd [hunk 15]
```diff
@@ -0,0 +1,540 @@
...
+      \(x) not_na(x),
+      Sepal.Length,
+      description = "Sepal.Length not_na"
+    ) %>%
+    validate_cols(
+      within_n_sds(3),
+      Sepal.Width,
+      description = "Sepal.Width outliers"
+    ) %>%
+    add_results(report)
```
Comment (jakubnowicki) [inline vignettes/targets_workflow.Rmd:457]: Please add a blank line below.

File: vignettes/targets_workflow.Rmd [hunk 16]
```diff
@@ -0,0 +1,540 @@
...
+
+error_localization <- function(report) {
+  report %>%
+    get_results(unnest = TRUE) %>%
+    filter(type == "error") %>%
+    select(index, column)
+}
+
+data_curation <- function(data, error_indexes) {
+  tar_assert_in("Sepal.Width", names(data))
```
Comment (jakubnowicki) [inline vignettes/targets_workflow.Rmd:469]: Please add a blank line below.

File: vignettes/targets_workflow.Rmd [hunk 17]
```diff
@@ -0,0 +1,540 @@
...
+  indexes_to_change <- error_indexes %>%
+    filter(column == "Sepal.Width") %>%
+    pull(index)
```
Comment (jakubnowicki) [inline vignettes/targets_workflow.Rmd:472]: Please add a blank line below.

File: vignettes/targets_workflow.Rmd [hunk 18]
```diff
@@ -0,0 +1,540 @@
...
+-   `columns_validator()` performs assertions on particular columns. It checks for uniqueness of the `index` column, that all values in `Species` column are in a given character vector, and that no outliers exist for `Sepal.Width` using a 3 standard deviation margin.
+
+-   `error_localization()` returns the precise location of the validation failures using the `data.validator::get_results()` function.
+
+-   `data_curation()` has input the error indexes and performs a simple naive replacement of all validation failures in the `Sepal.Width` column with the median.
+
+The `_targets.R` script looks like:
+
+```{r, eval=FALSE}
+# _targets.R file
```
Comment (jakubnowicki) [inline vignettes/targets_workflow.Rmd:489]: Please add a blank line below.

File: vignettes/targets_workflow.Rmd [hunk 19]
```diff
@@ -0,0 +1,540 @@
...
+library(targets)
```
Comment (jakubnowicki) [inline vignettes/targets_workflow.Rmd:490]: Please add a blank line below.

File: vignettes/targets_workflow.Rmd [hunk 20]
```diff
@@ -0,0 +1,540 @@
...
+source("R/functions.R")
```
Comment (jakubnowicki) [inline vignettes/targets_workflow.Rmd:491]: Please add a blank line below.

File: vignettes/targets_workflow.Rmd [hunk 21]
```diff
@@ -0,0 +1,540 @@
...
+options(crayon.enabled = FALSE)
+tar_option_set(
+  packages = c("readr", "dplyr", "data.validator", "assertr", "rlang",
+               "purrr")
+)
```
Comment (jakubnowicki) [inline vignettes/targets_workflow.Rmd:496]: Please add a blank line below.

**Review by jakubnowicki — COMMENTED** [review]
This is a really nicely written! Great job!

A few things to fix:
1. Please add comments to the code chunks that are not included, so it is easy to understand, what is happening there (a short sentence should be enough).
2. Please add blank lines to make code more readable.
3. I miss a summary section at the end of the tutorial - can you add something?

File: vignettes/targets_workflow.Rmd (hunk 1 above)
Comment (alexverse) [reply vignettes/targets_workflow.Rmd:130]: I agree

File: vignettes/targets_workflow.Rmd (hunk 3 above)
Comment (alexverse) [reply vignettes/targets_workflow.Rmd:254]: Added

File: vignettes/targets_workflow.Rmd (hunk 4 above)
Comment (alexverse) [reply vignettes/targets_workflow.Rmd:258]: Done!

File: vignettes/targets_workflow.Rmd (hunk 5 above)
Comment (alexverse) [reply vignettes/targets_workflow.Rmd]: I agree!

File: vignettes/targets_workflow.Rmd (hunk 17 above)
Comment (alexverse) [reply vignettes/targets_workflow.Rmd:589]: Added

File: vignettes/targets_workflow.Rmd (hunk 18 above)
Comment (alexverse) [reply vignettes/targets_workflow.Rmd:613]: Added

File: vignettes/targets_workflow.Rmd (hunk 19 above)
Comment (alexverse) [reply vignettes/targets_workflow.Rmd:615]: Done

File: vignettes/targets_workflow.Rmd (hunk 20 above)
Comment (alexverse) [reply vignettes/targets_workflow.Rmd:617]: Done

File: vignettes/targets_workflow.Rmd (hunk 21 above)
Comment (alexverse) [reply vignettes/targets_workflow.Rmd:622]: I agree!

**Review by jakubnowicki — APPROVED** [review]
Great job!

**Pull request merged**